File: compiler/src/feedback/items/AttributeFeedback.swift
func registerAttributeFeedback(into context: FeedbackContext) {
    context.register(
        id: "extra-attribute-value",
        query: Queries.fact(ExtraAttributeValue.self),
        rank: { _, _ in .syntax },
        location: { node, _ in (primary: node, secondary: []) },
    ) { writer, node, _ in
        writer.write(node)
        writer.write(" doesn't accept a value.")
        writer.writeBreak()
        writer.write("Try removing the value from this attribute.")
    }

    context.register(
        id: "duplicate-attribute",
        query: Queries.fact(DuplicateAttribute.self),
        rank: { _, _ in .syntax },
        location: { node, _ in (primary: node, secondary: []) },
    ) { writer, node, _ in
        writer.write(node)
        writer.write(" is defined more than once.")
        writer.writeBreak()
        writer.write("Try removing this attribute.")
    }

    context.register(
        id: "missing-attribute-value",
        query: Queries.fact(MissingAttributeValue.self),
        rank: { _, _ in .syntax },
        location: { node, _ in (primary: node, secondary: []) },
    ) { writer, node, _ in
        writer.write(node)
        writer.write(" is missing a value.")
        writer.writeBreak()
        writer.write("Try adding a value to this attribute using ")
        writer.write(code: ":")
        writer.write(".")
    }
}

File: compiler/feedback/src/items/attribute_feedback.rs
use crate::{FeedbackCtx, FeedbackRank};
use wipple_core::visit::attributes::{
    DuplicateAttribute, ExtraAttributeValue, MissingAttributeValue,
};
use wipple_queries::fact;

pub fn register(ctx: &mut FeedbackCtx<'_>) {
    ctx.feedback("extra-attribute-value")
        .query(fact::<ExtraAttributeValue>)
        .rank(|_| FeedbackRank::Syntax)
        .display(|_db, writer, node, _| {
            writer.node(node);
            writer.string(" doesn't accept a value.");
            writer.line_break();
            writer.string("Try removing the value from this attribute.");
        })
        .register();

    ctx.feedback("duplicate-attribute")
        .query(fact::<DuplicateAttribute>)
        .rank(|_| FeedbackRank::Syntax)
        .display(|_db, writer, node, _| {
            writer.node(node);
            writer.string(" is defined more than once.");
            writer.line_break();
            writer.string("Try removing this attribute.");
        })
        .register();

    ctx.feedback("missing-attribute-value")
        .query(fact::<MissingAttributeValue>)
        .rank(|_| FeedbackRank::Syntax)
        .display(|_db, writer, node, _| {
            writer.node(node);
            writer.string(" is missing a value.");
            writer.line_break();
            writer.string("Try adding a value to this attribute using ");
            writer.code(":");
            writer.string(".");
        })
        .register();
}
